File: src/ltd/rust_lang/utils/Stack.java
package ltd.rust_lang.utils;

import ltd.rust_lang.utils.list.DynamicArray;
import ltd.rust_lang.utils.list.MyList;

public class Stack<E> implements IStack<E> {
  private MyList<E> list = new DynamicArray<>();

  @Override
  public E pop() {
    return list.remove(size() - 1);
  }

  @Override
  public void push(E e) {
    list.add(e);
  }

  @Override
  public int size() {
    return list.size();
  }

  @Override
  public boolean isEmpty() {
    return list.isEmpty();
  }

  @Override
  public E top() {
    return list.get(size() - 1);
  }
}

File: src/main/kotlin/net/wu_chinese/utils/stack/Stack.kt
package net.wu_chinese.utils.stack

import net.wu_chinese.utils.list.Vec

class Stack<E> : IStack<E> {
    private val vec = Vec<E>()

    override fun pop(): E? {
        return vec.remove(size() - 1)
    }

    override fun push(e: E?) {
        vec.add(e)
    }

    override fun size(): Int {
        return vec.size()
    }

    override val isEmpty: Boolean
        get() = vec.isEmpty

    override fun peek(): E? {
        return vec[size() - 1]
    }

    override fun clear() {
        vec.clear()
    }
}
